### PR TITLE
:bug: Fix coltypes for projects with dags

### LINKED ISSUE #467: Unexpected Behavior: redcap_metadata_coltypes assumes that "record_id" field is an integer
*!! Note about security !! If your question/issue involves sensitive information, please do not post it in this public location.  You might assist someone in compromising your institution's server or network.  The [REDCap Community](https://community.projectredcap.org/spaces/20/index.html) is the appropriate forum.  Please ask your institution's REDCap administrator to post a question on this site, with the tags [api](https://community.projectredcap.org/topics/337/api.html) and [redcapr](https://community.projectredcap.org/topics/23921/redcapr.html).*

**Describe the behavior**: Please provide a clear and concise description of the scenario and the behavior.  *Be careful not to include tokens, PHI (protected health information), or other information that should not be public!*

First off, I ❤️ the `redcap_metadata_coltypes` function, such a hidden gem of this package. It's really saving my bacon in this ETL I'm working on across several very large/old/complex REDCap projects.

In working with it, I discovered that it just assumes that, if autonnumbering is enabled, the `record_id` field (or first field of the project) is going to be an integer, which doesn't work when DAGs are involved (since records look more like this: `5373-100` than this `100`)

**Expected behavior**: A clear and concise description of what you expected to happen.

It would be great if this function were "DAG" aware, and looked for that **first** before looking at autonumbering, in order to determine the data type of the `record_id` field.

**Desktop (please complete the following information):**
 - OS: Mac
 - REDCap version: 12.5.15
 - REDCapR Version: GitHub development version

**Additional context**: Add any other information about the specific problem, such as a data dictionary or screenshot.

Here's is the part of the code that is causing the issue for me:

https://github.com/OuhscBbmc/REDCapR/blob/0b098f4f7ac3f7e51277b8e12ea2b6f2e0bde238/R/redcap-metadata-coltypes.R#L241

https://github.com/OuhscBbmc/REDCapR/blob/0b098f4f7ac3f7e51277b8e12ea2b6f2e0bde238/R/redcap-metadata-coltypes.R#L363-L366

I don't think the change would be that difficult 🤞🏻 and I'm happy to contribute if you're aligned with my approach

### PR DESCRIPTION
**Pull request by pwildenhain**
Fixes #467

### CODE REVIEW

File: R/redcap-metadata-coltypes.R [hunk 1]
```diff
@@ -371,12 +374,13 @@ redcap_metadata_internal <- function(
...
     d_meta %>%
     dplyr::mutate(
+      dags        = (.dags & (.data$field_name == .record_field)),
       autonumber  = (.autonumber & (.data$field_name == .record_field)),
     ) %>%
     dplyr::mutate(
       response =
         dplyr::case_when(
-          autonumber                                          ~ paste0("col_integer()"                        , "~~record_autonumbering is enabled for the project"),
+          autonumber & !dags                                  ~ paste0("col_integer()"                        , "~~record_autonumbering is enabled for the project"),
```
Comment (wibeasley) [inline R/redcap-metadata-coltypes.R:383]: @pwildenhain, how do you feel about a line here that spells out the combo (of dag & autonumbering) suggests a character?

```r
autonumber &  dags ~ paste0("col_character()" , "~~record_autonumbering & DAGs are enabled for the project"),
```

**Review by wibeasley — COMMENTED** [review]
@pwildenhain, can you please add an entry in https://github.com/OuhscBbmc/REDCapR/blob/main/NEWS.md#minor-enhancements?

Maybe along the lines of "`redcap_meta_coltypes()` correctly determines data type for autonumber fields.  It suggests a character if the project has DAGs, and an integer if not. (@pwildenhain , #472)"

**Comment by wibeasley** [comment]
@pwildenhain, I see you just added me as a reviewer.  I think you've added everything except for the line that [explicitly specifies "character"](https://github.com/OuhscBbmc/REDCapR/pull/472#discussion_r1119316750) for autonumbering w/ dags.  Am I overlooking it?

As the waterfall is written now, it shouldn't affect the recommended type.  But two benefits are
1. an explicit explanatory comment for the humans to read and
2. more robustness (like I mess up the waterfall later, or something goes weird like a user is able to add a date validation on the field before it became an autonumber)

**Comment by wibeasley** [comment]
yeah, that's a good way to say it.

Ideally the col_types entry has an informative comment/explanation.

If it helps to visualize, this is [an example current output](https://ouhscbbmc.github.io/REDCapR/reference/redcap_metadata_coltypes.html): 

```r
# A simple project
token      <- "9A81268476645C4E5F03428B8AC3AA7B" # 153
col_types  <- redcap_metadata_coltypes(uri, token)
#> # col_types <- readr::cols_only( # Use `readr::cols_only()` to restrict the retrieval to only these columns
#> col_types <- readr::cols( # Use `readr::cols()` to include unspecified columns
#>   # [field]                       [readr col_type]            [explanation for col_type]
#>   record_id                     = readr::col_integer()    , # record_autonumbering is enabled for the project
#>   name_first                    = readr::col_character()  , # field_type is text and validation isn't set
#>   name_last                     = readr::col_character()  , # field_type is text and validation isn't set
#>   address                       = readr::col_character()  , # field_type is note
#>   telephone                     = readr::col_character()  , # validation is 'phone'
#>   email                         = readr::col_character()  , # validation is 'email'
#>   dob                           = readr::col_date()       , # validation is 'date_ymd'
#>   age                           = readr::col_character()  , # field_type is text and validation isn't set
#>   sex                           = readr::col_character()  , # field_type is radio
...
```

**Comment by pwildenhain** [comment]
Hold off on merging this for now -- I'm just now realizing that this function will fail if the user is currently assigned to a DAG -- since you can't access the "Export DAGs" API endpoint when you assigned to a DAG. I think we can maneuver around this though.

**Comment by pwildenhain** [comment]
Ok @wibeasley I've added some code to account for that edge case, and I've tested it with one of my projects to make sure it works. PR is ready for review again.

**Comment by pwildenhain** [comment]
Pinging @wibeasley again to review -- not sure why the tests failed

**Comment by wibeasley** [comment]
I meant to check it out more deeply.  I thought it was something fluky and that's why I hit some limit of four rerunning.  I've changed the base to dev and I'll try some things now.

File: R/redcap-metadata-coltypes.R [hunk 2]
```diff
@@ -380,7 +380,8 @@ redcap_metadata_internal <- function(
     dplyr::mutate(
       response =
         dplyr::case_when(
-          autonumber & !dags                                  ~ paste0("col_integer()"                        , "~~record_autonumbering is enabled for a project without DAGs"),
+          dags                                                ~ paste0("col_character()"                      , "~~DAGs are enabled for the project"),
```
[the remark this answers is not in the document]
Comment (pwildenhain) [reply R/redcap-metadata-coltypes.R:383]: No worries haha. `dags` will only evaluate to `TRUE` when the field is the `.record_id` field AND the project has DAGs (similar to how `autonumber` already works

https://github.com/pwildenhain/REDCapR/blob/ef2171960807ede2136172e59b866953b7928666/R/redcap-metadata-coltypes.R#L375-L379
